database/api.go

```diff
@@ -0,0 +1,172 @@
+//go:generate mockgen -source=$GOFILE -package $GOPACKAGE -destination api.mock.gen.go
+package database
+
+import (
+	"github.com/beaconsoftwarellc/gadget/v2/database/qb"
+	"github.com/jmoiron/sqlx"
+)
+
+// Database is a reduced database interface which only exposes the transaction and select functions
+type API interface {
+	// Begin starts a transaction
+	Begin() error
+	// Commit commits the transaction
+	Commit() error
+	// Rollback aborts the transaction
+	Rollback() error
+
+	// Create initializes a Record and inserts it into the Database
+	Create(obj Record) error
+	// Read populates a Record from the database
+	Read(obj Record, pk PrimaryKeyValue) error
+	// ReadOneWhere populates a Record from a custom where clause
+	ReadOneWhere(obj Record, condition *qb.ConditionExpression) error
+	// Select executes a given select query and populates the target
+	Select(target interface{}, query *qb.SelectQuery) error
+	// SelectList of Records into target based upon the passed query
+	SelectList(target interface{}, query *qb.SelectQuery, options *ListOptions) error
+	// ListWhere populates target with a list of records from the database
+	ListWhere(meta Record, target interface{}, condition *qb.ConditionExpression, options *ListOptions) error
+	// Update replaces an entry in the database for the Record using a transaction
+	Update(obj Record) error
+	// UpdateWhere updates fields for the Record based on a supplied where clause
+	UpdateWhere(obj Record, where *qb.ConditionExpression, fields ...qb.FieldValue) (int64, error)
+	// Delete removes a row from the database
+	Delete(obj Record) error
+	// DeleteWhereTx removes row(s) from the database based on a supplied where clause in a transaction
+	DeleteWhere(obj Record, condition *qb.ConditionExpression) error
+}
+
+// New
+func NewAPI(db *Database, tx *sqlx.Tx) API {
+	return &dbapi{Tx: tx, Database: db}
+}
+
+type dbapi struct {
+	*sqlx.Tx
+	*Database
+}
+
+// Begin starts a transaction
+func (d *dbapi) Begin() error {
+	var err error
+
+	if d.Tx == nil {
+		d.Tx, err = d.Beginx()
+	}
+
+	return err
+}
+
+// Rollback aborts the transaction
+func (d *dbapi) Rollback() error {
+	if d.Tx != nil {
+		defer d.cleanTx()
+		return d.Tx.Rollback()
+	}
+
+	return nil
```

Comment:
calling rollback when there is no transaction should be an error. Also, let's not use a defer just to nil out the tx on the next line.